fs/nfsd/nfs4acl.c

```diff
@@ -751,26 +908,160 @@ static int nfs4_acl_nfsv4_to_posix(struct nfs4_acl *acl,
 	return ret;
 }
 
-__be32 nfsd4_acl_to_attr(enum nfs_ftype4 type, struct nfs4_acl *acl,
-			 struct nfsd_attrs *attr)
+static __be32
+nfsd4_acl_to_attr_posix(enum nfs_ftype4 type, struct nfs4_acl *acl,
+			fsacl_t *fsaclp)
 {
 	int host_error;
 	unsigned int flags = 0;
 
 	if (!acl)
 		return nfs_ok;
-
 	if (type == NF4DIR)
 		flags = NFS4_ACL_DIR;
 
-	host_error = nfs4_acl_nfsv4_to_posix(acl, &attr->na_pacl,
-					     &attr->na_dpacl, flags);
+	host_error = nfs4_acl_nfsv4_to_posix(acl, &fsaclp->posixacl.na_pacl,
+					     &fsaclp->posixacl.na_dpacl, flags);
 	if (host_error == -EINVAL)
 		return nfserr_attrnotsupp;
 	else
 		return nfserrno(host_error);
 }
 
+static int
+convert_ace_to_nfs41(u32 *p, const struct nfs4_ace *ace)
+{
+	int error = 0;
+	nfsace4i nace;
+
+	nace = (nfsace4i) {
+		.type = (acetype4)ace->type,
+		.flag = (aceflag4)ace->flag & NFS41_FLAGS,
+		.access_mask = ace->access_mask & NFS4_ACE_MASK_ALL,
+		.who = -1,
+	};
+
+	/* Audit and Alarm are not currently supported */
+	if (nace.type > NFS4_ACE_ACCESS_DENIED_ACE_TYPE)
+		return -EINVAL;
+
+	switch (ace->whotype) {
+	case NFS4_ACL_WHO_OWNER:
+		nace.iflag = ACEI4_SPECIAL_WHO;
+		nace.who = ACE4_SPECIAL_OWNER;
+		break;
+	case NFS4_ACL_WHO_GROUP:
+		nace.iflag = ACEI4_SPECIAL_WHO;
+		nace.who = ACE4_SPECIAL_GROUP;
+		break;
+	case NFS4_ACL_WHO_EVERYONE:
+		nace.iflag = ACEI4_SPECIAL_WHO;
+		nace.who = ACE4_SPECIAL_EVERYONE;
+		break;
+	case NFS4_ACL_WHO_NAMED:
+		if (ace->flag & NFS4_ACE_IDENTIFIER_GROUP)
+			nace.who = (u_int)__kgid_val(ace->who_gid);
+		else
+			nace.who = (u_int)__kuid_val(ace->who_uid);
+
+		BUG_ON((nace.who == -1) || (nace.who > U32_MAX));
```

Comment:
I am not getting this condition.  Check for -1 may not work after conversion to different unsigned type.  Also I don't see how __kgid_val/__kuid_val  ever return -1.  Sense of U32_MAX after conversion to u_int is also questionable.  And do we want any BUG_ON on user-supplied data?